sdk/src/entrypoint.rs

```diff
@@ -16,16 +16,82 @@ use std::{
+    /// CustomError allows on-chain programs to implement program-specific error types and see
+    /// them returned by the Solana runtime. A CustomError may be any type that is represented
+    /// as or serialized to a u32 integer.
+    ///
+    /// NOTE: u64 requires special serialization to avoid the loss of precision in JS clients and
+    /// so is not used for now.
+    CustomError(u32),
+
+    /// The arguments provided to a program instruction where invalid
+    InvalidArgument,
+
+    /// An instruction's data contents was invalid
+    InvalidInstructionData,
+
+    /// An account's data contents was invalid
+    InvalidAccountData,
+
+    /// An account's data was too small
+    AccountDataTooSmall,
+
+    /// An account's balance was too small to complete the instruction
+    InsufficientFunds,
+
+    /// The account did not have the expected program id
+    IncorrectProgramId,
+
+    /// A signature was required but not found
+    MissingRequiredSignature,
+
+    /// An initialize instruction was sent to an account that has already been initialized.
+    AccountAlreadyInitialized,
+
+    /// An attempt to operate on an account that hasn't been initialized.
+    UninitializedAccount,
+
+    /// The instruction expected additional account keys
+    NotEnoughAccountKeys,
+
+    /// Failed to borrow a reference to an account, already borrowed
+    AccountBorrowFailed,
+}
+
+impl From<InstructionError> for u32 {
+    fn from(error: InstructionError) -> Self {
+        match error {
+            InstructionError::InvalidArgument => std::u32::MAX - 1,
+            InstructionError::InvalidInstructionData => std::u32::MAX - 2,
+            InstructionError::InvalidAccountData => std::u32::MAX - 3,
+            InstructionError::AccountDataTooSmall => std::u32::MAX - 4,
+            InstructionError::InsufficientFunds => std::u32::MAX - 5,
+            InstructionError::IncorrectProgramId => std::u32::MAX - 6,
+            InstructionError::MissingRequiredSignature => std::u32::MAX - 7,
+            InstructionError::AccountAlreadyInitialized => std::u32::MAX - 8,
+            InstructionError::UninitializedAccount => std::u32::MAX - 9,
+            InstructionError::NotEnoughAccountKeys => std::u32::MAX - 10,
+            InstructionError::AccountBorrowFailed => std::u32::MAX - 11,
+            InstructionError::CustomError(error) => error,
+        }
+    }
+}
+
 /// User implemented program entrypoint
 ///
 /// program_id: Program ID of the currently executing program
 /// accounts: Accounts passed as part of the instruction
 /// data: Instruction data
-pub type ProcessInstruction =
-    fn(program_id: &Pubkey, accounts: &mut [AccountInfo], instruction_data: &[u8]) -> u32;
+pub type ProcessInstruction = fn(
+    program_id: &Pubkey,
+    accounts: &mut [AccountInfo],
+    instruction_data: &[u8],
+) -> Result<(), InstructionError>;
 
 /// Programs indicate success with a return value of 0
-pub const SUCCESS: u32 = 0;
```

Comment:
Why can't we keep SUCCESS == 0?

Reply:
We could, but if folks define their own error enums they will by default start at zero and we would have to bump a "0" error up to something else so it doesn't look like a success.  This hidden bump can be confusing

Reply:
Ah, I see.   Custom errors always have the 31st bit set maybe?   0 == SUCCESS is pretty nice

Reply:
hehe, this is why I'm throwing it out for thoughts, I like 0 as SUCCESS too.  we could set the 31st bit but that still isn't super nice for developers.    We need to also support built-in InstructionErrors, I guess those could be down low (1, 2, 3, etc...)

Reply:
Maybe go i32?  The user gets the negatives, we get the positives.

Reply:
Or catch any conflicting `CustomError` conflicting values like "0" being returned, report it, and translate it to something else.  For example 0 == success, high order bit error messages are built-ins like `InstructionError::InvalidArgument` and if `process_instruction` returns a `CustomError(0)` then we report that error "0" was returned but then translate to `InstructionError::ConflictingErrorCode` or maybe something more specific.  On the C side, we already have 0 defined as success, we go ahead and define the other return values and leave it up to the user not to conflict.